### PR TITLE
Add subscription model

### PR DESCRIPTION
**Pull request by thomasleese**
This adds a subscription model which is used to link together the subscriber and the subscriber list classes. 

[Trello Card](https://trello.com/c/JmOA1348/256-subscription-model)

### CODE REVIEW

File: app/models/subscriber.rb [hunk 1]
```diff
@@ -2,4 +2,6 @@ class Subscriber < ActiveRecord::Base
   validates :address, presence: true
   validates :address, format: { with: /@/, message: "is not an email address" }
   validates :address, uniqueness: true
+
+  has_many :subscribers, dependent: :destroy
```
Comment (gpeng) [inline app/models/subscriber.rb]: Should this be `subscriptions`?

File: app/models/subscriber_list.rb [hunk 2]
```diff
@@ -9,6 +9,8 @@ class SubscriberList < ApplicationRecord
   validate :tag_values_are_valid
   validate :link_values_are_valid
 
+  has_many :subscribers, dependent: :destroy
```
Comment (gpeng) [inline app/models/subscriber_list.rb]: And `subscriptions` here too?

File: app/models/subscriber.rb (hunk 1 above)
Comment (thomasleese) [reply app/models/subscriber.rb]: Yep!

**Comment by thomasleese** [comment]
@gpeng Regarding the `dependent: :destroy` I figured it might be useful between subscriptions and subscribers to easily delete subscriptions, but I left it out on the subscriber lists because I imagine we might want to be absolutely sure about only deleting subscriber lists when they have no subscribers.

**Comment by thomasleese** [comment]
It didn't occur to me to include both the `has_many` and the `has_many :through` associations in each model, which makes a lot of sense, so I've added that now.

File: app/models/subscriber.rb [hunk 3]
```diff
@@ -2,4 +2,7 @@ class Subscriber < ActiveRecord::Base
   validates :address, presence: true
   validates :address, format: { with: /@/, message: "is not an email address" }
   validates :address, uniqueness: true
+
+  has_many :subscriptions, dependent: :destroy
+  has_many :subscription_lists, through: :subscriptions
```
Comment (gpeng) [inline app/models/subscriber.rb]: `subscriber_lists` here

Comment (thomasleese) [reply app/models/subscriber.rb]: Clearly struggling with words this morning...

**Comment by thomasleese** [comment]
Thanks @gpeng, good point. I've added some tests now that should cover everything.

**Comment by gpeng** [comment]
Actually... Should we also add the uniqueness constraint as a validation in the model so we get `ActiveRecord` errors before the database ones kick in?

**Comment by thomasleese** [comment]
@gpeng Good idea, have added one if you could take a look.

**Pull request merged**